language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/builder/FunctionCompletionItemBuilder.java

```diff
@@ -373,6 +379,40 @@ private static String getQualifiedFunctionName(String functionName, BallerinaCom
         return modulePrefix + SyntaxKind.COLON_TOKEN.stringValue() + functionName;
     }
 
+    /**
+     * Creates and returns the main function completion item.
+     *
+     * @param context   Ballerina completion context
+     * @return {@link CompletionItem} generated main function completion item.
+     */
+    public static CompletionItem buildMainFunction(BallerinaCompletionContext context) {
+        NonTerminalNode node = context.getNodeAtCursor();
+        Optional<Token> lastQualifier = Optional.empty();
+        while (node != null) {
+            lastQualifier = ModulePartNodeContextUtil.getLastQualifier(context, node);
+            if (lastQualifier.isPresent() || node.kind() == SyntaxKind.MODULE_PART) {
+                break;
+            }
+            node = node.parent();
+        }
```

Comment:
Shall we embed the parent visiting logic inside the  ModulePartNodeContextUtil.getLastQualifier method. Because ideally, we should not have to visit the parents to get the last qualifier. Are there any breaking cases if we do so?

Reply:
The module part context is very unpredictive and I think that's why we had to add this logic. @mindula can add why we needed the parent visit thing. I don't think we should make the parent visit thing common by moving to the util

Reply:
+1 to having the logic as is then

Reply:
There are some cases where we need to visit the parents to get the last qualifier. For example, consider the following case,
```
public <cursor>
function name() {

}
```
when we try to get the last qualifier using ModulePartNodeContextUtil.getLastQualifier method for the above cursor position, it gives Optional.empty()